### ISSUE TITLE
Update S4210: Rule should not suggest STAThreadAttribute for async main

### ISSUE THREAD

**Issue by nnpcYvIVl** [SonarSource/sonar-dotnet#2352]
S4210 is flagged on ```Main``` on an async WinForms program but even if I add the ```STAThread``` attribute the ApartmentState is actually MTA. Is S4210 a false positive for async mains?

```csharp
using System.Threading;
using System.Threading.Tasks;

namespace S4210_False_Positive
{
    static class Program
    {
        static async Task Main()
        {
            System.Windows.Forms.MessageBox.Show(Thread.CurrentThread.GetApartmentState().ToString());
            await Task.CompletedTask;
        }
    }
}
```

### Related information

* SonarC# Version: Latest
* Visual Studio Version: 15.9.9

**Comment by Evangelink** [comment]
Hi @nnpcYvIVl,

I am not entirely sure to understand your question/comment but as far as I know, `STAThread` is the only correct/supported way to have a main method in the context of a WinForm application.

**Comment by nnpcYvIVl** [comment]
Sorry for the confusion. Does this clear it up?

A WinForms void Main() runs in an MTA ApartmentState.

```csharp
static void Main()
{
    System.Windows.Forms.MessageBox.Show(Thread.CurrentThread.GetApartmentState().ToString());
}
```

Adding the [STAThread] attribute forces it to run in an STA ApartmentState.

```csharp
[STAThread]
static void Main()
{
    System.Windows.Forms.MessageBox.Show(Thread.CurrentThread.GetApartmentState().ToString());
}
```

A WinForms async Task Main() also runs in an MTA ApartmentState.

```csharp
static async Task Main()
{
    System.Windows.Forms.MessageBox.Show(Thread.CurrentThread.GetApartmentState().ToString());
}
```

Adding the [STAThread] attribute doesn't do anything. It runs in an MTA ApartmentState.

```csharp
[STAThread]
static async Task Main()
{
    System.Windows.Forms.MessageBox.Show(Thread.CurrentThread.GetApartmentState().ToString());
}
```

S4210 is recommending adding the STAThread attribute on an async Main even though it has no effect.

**Comment by Evangelink** [comment]
Hi @nnpcYvIVl,

This is clear! Thank you.

FYI I am linking a but that was reported on Roslyn https://github.com/dotnet/roslyn/issues/22112 about the fact that `STAThread` is not passed through with async main methods. And here is the link to the C# language spec discussion about it https://github.com/dotnet/csharplang/issues/97

So to answer your question, yes this is a FP and we should also have another rule that will actually raise an issue when you use the `[STAThread]` attribute on an async main method.
